sw/airborne/filters/pid.h

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (C) 2018 Gautier Hattenberger
+ *
+ * This file is part of paparazzi.
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+/** @file filters/pid.h
+ *  @brief Several forms of PID controllers
+ *
+ */
+
+#ifndef PID_H
+#define PID_H
+
+/** Simple PID structure
+ *  floating point
+ *
+ * u_k = Kp * e_k + Kd * (e_k - e_k-1) / dt + Ki * (sum (e_k * dt))
+ *
+ * with:
+ *  u = outputs
+ *  e = inputs
+ *  Kp = proportional gain
+ *  Kd = derivative gain
+ *  Ki = integral gain
+ *  dt = time since last input
+ */
+struct PID_f {
+  float u;        ///< output
+  float e[2];     ///< input
+  float sum;      ///< integral of input
+  float g[3];     ///< controller gains (Kp, Kd, Ki)
+  float max_sum;  ///< windup protection, max of Ki * sum(e_k * dt)
+};
+
+static inline void init_pid_f(struct PID_f *pid, float Kp, float Kd, float Ki, float max_sum)
+{
+  *pid = (struct PID_f) {
+    0.f,
+    { 0.f , 0.f },
+    0.f,
+    { Kp, Kd, Ki },
+    max_sum
+  };
+}
+
+/** Update PID with a new value and return new command.
+ *
+ * @param pid pointer to PID structure
+ * @param value new input value of the PID
+ * @param dt time since last input (in seconds)
+ * @return new output command
+ */
+static inline float update_pid_f(struct PID_f *pid, float value, float dt)
+{
+  pid->e[1] = pid->e[0];
+  pid->e[0] = value;
+  float integral = pid->g[2] * (pid->sum + value);
+  if (integral > pid->max_sum) {
+    integral = pid->max_sum;
+  } else if (integral < -pid->max_sum) {
+    integral = -pid->max_sum;
+  } else {
+    pid->sum += value;
+  }
+  pid->u = pid->g[0] * pid->e[0] + pid->g[1] * (pid->e[0] - pid->e[1]) / dt + integral;
+  return pid->u;
+}
+
+/** Get current value of the PID command.
+ *
+ * @param pid pointer to PID structure
+ * @return current value of PID command
+ */
+static inline float get_pid_f(struct PID_f *pid)
+{
+  return pid->u;
+}
+
+/** Reset PID struture, gains left unchanged.
+ *
+ * @param pid pointer to PID structure
+ */
+static inline void reset_pid_f(struct PID_f *pid)
+{
+  pid->u = 0.f;
+  pid->e[0] = 0.f;
+  pid->e[1] = 0.f;
+  pid->sum = 0.f;
+}
+
```

Comment:
I like this tool :P

What about a function for resetting only the integral part (sum)? Or better if you can set the value of "sum" to whatever value you wish, i.e., its initial condition. (I believe Simulink has it too).

Reply:
If you only need to reset/set the integral, maybe you can do it directly by changing the `sum` field ? Do you really think it is worth a dedicated function ? (probably better I guess...)

Reply:
Not really. I think setting the field "sum" directly should be fine although not elegant :P